### PR TITLE
[UserAPI] [UserIndicator Plugin] [ExampleUserProvider Plugin] New user api and related functionality

### LINKED ISSUE #4407: Implement User API
Introduce user "awareness" into Open MCT to enable user and role specific functionality such as role-specific home pages.

The specific details of how users are identified and authenticated will be platform specific, so we need an API that delegates those details to a "User Provider".

Here are some thoughts on how we can break this problem down:
1. Introduce a new "User API" eg. `openmct.user.getCurrentUser()`
2. Add a new plugin that adds an indicator that shows the name of the currently logged in user. This plugin would be installed by default, but should only show an indicator if a user provider is defined in the API.
3. Add an example User Provider Plugin that exercises the new User API. This plugin could add a dialog that allows the user to identify themselves, and then displays the provided name in the indicator defined in step 2. This plugin will not be installed by default, and will just be provided as an example to developers

A separate issue will be created to integrate the new Open MCT User API with Yamcs' user API.

I have some initial thoughts on what the user API might look like, which we can iterate on. One key thing to note is that I *think* we only want to support a single user provider at a time? Unlike view, object, and telemetry providers where we need to support many.

``` JavaScript
// This is how the current user is checked (eg. from an indicator that displays the logged in user)
let user = openmct.user.getCurrentUser(); // Returns a promise that resolves with a user or undefined

// This is how the `getCurrentUser()` function might be implemented:
openmct.user.getCurrentUser = function() {
    if (!this.provider || this.provider.isLoggedIn) {
        return Promise.resolve(undefined);
    } else {
        return this.provider.getCurrentUser();
    }
}

// This is how a new user provider is registered
openmct.user.setUserProvider({
    isLoggedIn() {

    }
    async getCurrentUser() {
        return {
            id: 'some-user-id',
            fullName: '',
            homeUrl: '' // MAY be built dynamically, based on role.
        }
    }
    async hasRole(roleId) {
    }
});
```

### PR DESCRIPTION
**Pull request by jvigliotta**
### All Submissions:

closes #4407

Adds a new user api, a user indicator plugin which will display if a user provider is set and an example user provider to use for testing and as an example for developers.

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [x] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [ ] Code style and in-line documentation are appropriate?
* [ ] Commit messages meet standards?
* [ ] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Review by akhenry — CHANGES_REQUESTED** [review]
This looks really great, thanks.

On reflection, I don't think we want to deal with login and logout events. Either the application does not support users (there are no providers defined), or it does. 

In the case that users are supported, a user is always logged in (although determining the details of a user is asynchronous, so still requires a promise).

My reasoning for this is:
* If a user can be logged in or out, we have to handle state change events, which adds a lot of complexity to our user-aware views.
* What does it really mean to be logged out in a system that supports users. What can a logged out user do? Do our views really want to deal with that ambiguity? In the scenario where a system really wanted to support logged out users in some sort of a 'guest user' state, then they can define a guest user and role, and reload the application logged in as that user.

File: src/api/user/UserAPI.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+import User from './User';
+
+class UserAPI extends EventEmitter {
+    constructor(openmct) {
+        super();
+
+        this._openmct = openmct;
+        this._provider = undefined;
+
+        this.User = User;
```
Comment (akhenry) [inline src/api/user/UserAPI.js:37]: 👍 Ah, good catch, it can't be static, it has to be on the instance to be usable as `openmct.user.User`

**Review by akhenry — DISMISSED** [review]
Beautiful! Love it! _Excellent_ tests, nice work.

### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y++
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue) Y

File: src/plugins/userIndicator/pluginSpec.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    afterEach(() => {
+        return resetApplicationState(openmct);
+    });
+
+    it('will not show, if there is no user provider', () => {
+        userIndicator = openmct.indicators.indicatorObjects
+            .find(indicator => indicator.key === 'user-indicator');
+
+        const userIndicatorMissing = userIndicator === null || userIndicator === undefined;
```
Comment (davetsay) [inline src/plugins/userIndicator/pluginSpec.js:64]: do we set the user indicator to null ever?

File: src/plugins/userIndicator/pluginSpec.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        expect(userIndicatorMissing).toBe(true);
+    });
+
+    describe('with a user provider installed', () => {
+
+        beforeEach(() => {
+            provider = new ExampleUserProvider(openmct);
+            spyOn(provider, '_login');
```
Comment (davetsay) [inline src/plugins/userIndicator/pluginSpec.js:72]: is it possible to mock this behavior without spying on an internal?

File: src/api/user/UserAPISpec.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    afterEach(() => {
+        return resetApplicationState(openmct);
+    });
+
+    describe('with regard to user providers', () => {
+
+        it('allows you to specify a user provider', () => {
+            openmct.user.setProvider(new ExampleUserProvider(openmct));
+
+            expect(openmct.user._provider).toBeInstanceOf(ExampleUserProvider);
```
Comment (davetsay) [inline src/api/user/UserAPISpec.js:51]: can we test this without knowing about internals?

It looks like what you've done in the example spec is what should be done here. Testing that providers don't exist, and then they do when added. And if the provider needs to be checked, the check example spec has of `isInstanceOf` should do the trick.

**Review by davetsay — DISMISSED** [review]
couple minor details/questions.

This is awesome work.

File: src/plugins/userIndicator/pluginSpec.js (hunk 2 above)
Comment (jvigliotta) [reply src/plugins/userIndicator/pluginSpec.js:64]: No clue. I pulled this from another indicator plugin.

File: src/plugins/exampleUser/plugin.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*****************************************************************************
```
Comment (nikhilmandlik) [inline src/plugins/exampleUser/plugin.js:1]: I think we should move this to `/openmct/example` folder

File: src/plugins/userIndicator/components/UserIndicator.vue [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ Open MCT includes source code licensed under additional open source
+ licenses. See the Open Source Licenses file (LICENSES.md) included with
+ this source code distribution or the Licensing information page available
+ at runtime from the About dialog for additional information.
+-->
+
+<template>
+<div class="c-indicator icon-person no-minify c-indicator--clickable">
+    <span class="label c-indicator__label">
+        {{ name }}
```
Comment (nikhilmandlik) [inline src/plugins/userIndicator/components/UserIndicator.vue:26]: wondering we need to show user info on click indicator?

Comment (jvigliotta) [reply src/plugins/userIndicator/components/UserIndicator.vue:26]: Good question, maybe one for @charlesh88 .  He actually need to take a look at this one. He may have some thoughts.

File: src/plugins/exampleUser/plugin.js (hunk 5 above)
Comment (akhenry) [reply src/plugins/exampleUser/plugin.js:1]: Good catch, I agree. It's not ever intended for production usage.

File: src/plugins/plugins.js [hunk 7]
```diff
@@ -240,6 +244,8 @@ define([
     plugins.Clock = Clock.default;
     plugins.Timer = Timer.default;
     plugins.DeviceClassifier = DeviceClassifier.default;
+    plugins.UserIndicator = UserIndicator.default;
+    plugins.ExampleUser = ExampleUser.default;
```
Comment (akhenry) [inline src/plugins/plugins.js:248]: I think maybe we should group our example plugins. eg:
``` JavaScript
plugins.example = {};
plugins.example.User = ExampleUser.default
```

File: src/plugins/userIndicator/pluginSpec.js (hunk 2 above)
Comment (akhenry) [reply src/plugins/userIndicator/pluginSpec.js:64]: Good question, null seems unnecessary here? I don't think `.find` could possibly return null in this case.

File: src/plugins/userIndicator/pluginSpec.js (hunk 3 above)
Comment (akhenry) [reply src/plugins/userIndicator/pluginSpec.js:72]: Agreed, that would be ideal.

**Review by akhenry — CHANGES_REQUESTED** [review]

**Comment by charlesh88** [comment]
Pushed changes to markup and config:

image

I stubbed in Login/Logout buttons in the indicator and removed the `no-minify`:

image

### Questions
- Why aren't we using the Open MCT form component? This approach doesn't give us access to the validation indicators like
image
image

- The Login button should only be enabled when the form is valid - again, available in the standard form component. 
- No password field? Wasn't in markup and didn't see a provision to dynamically insert. I stubbed it in, but commented out in ExampleUserProvider.js.
- Invalid form messaging shouldn't be being done via a Notification message - it appears beneath the overlay blocker and won't be seen. Instead, we should use built-in error messaging as in the Time Conductor. I'd include a screenshot here to illustrate, but it's not working right now...
- Do we get configs where the user can access the indicator without being logged in? If not, then the indicator shouldn't show at all until a user is logged in. Clicking logout would immediately go to the Login dialog.

**Comment by akhenry** [comment]
@charlesh88 I can address some of these questions. Open MCT will not handle login/logout, it will be handled by an external system that we have no control over, so we don't need to support login or logout behavior. In fact, we can safely assume that if the application is loaded successfully, then the user has been authenticated. The User API just gives us a way of interrogating an external system to find out _who_ the user is, not to handle login or authentication.

The Example User Provider that @jvigliotta has defined is provided purely as an example of how to use the User API. To do that, we need a username at least, so @jvigliotta has defined a simple form for collecting a user name. It's not intended as a general-purpose login form, so we don't _really_ need a password field. 

I agree that if it's expedient to use the forms API for this (it wasn't merged when this work was done), then that would be ideal, but we shouldn't spend a ton of time on this. @jvigliotta Would this be a lot of work?

**Comment by jvigliotta** [comment]
@akhenry It doesn't look like it'd be a lot of work. I'll sub in using the form API.

**Comment by charlesh88** [comment]
> @akhenry It doesn't look like it'd be a lot of work. I'll sub in using the form API.

Thanks for the details gents, this all makes sense. Totally agree on not wanting to add a lot of work here.

**Comment by jvigliotta** [comment]
@charlesh88 when I'm making the changes I need to make, would you like me to remove the login/logout stuff?

File: src/plugins/userIndicator/pluginSpec.js (hunk 3 above)
Comment (jvigliotta) [reply src/plugins/userIndicator/pluginSpec.js]: Done and done, I added an auto login feature that let's you bypass the login form while specifying a username.

File: src/plugins/plugins.js (hunk 7 above)
Comment (jvigliotta) [reply src/plugins/plugins.js:248]: I'll move this!

**Comment by unlikelyzero** [comment]
### Acceptance Tests

- [ ] Username appears in the title and reflects NDC Username once authenticated
- [ ] Username persists on refresh
- [ ] Username appears in multiple tabs and multiple browsers
- [ ] Username does not appear in title if canceling the Authentication method. (Note: this negative case results in console errors.
- [ ]

File: src/api/forms/components/FormProperties.vue [hunk 8]
```diff
@@ -105,6 +105,20 @@ export default {
                 .some(([key, value]) => {
                     return value;
                 });
+        },
+        submitLabel() {
+            if (this.model.buttons && this.model.buttons.submit && this.model.buttons.submit.label) {
```
Comment (akhenry) [inline src/api/forms/components/FormProperties.vue:110]: This should be split into multiple lines.

File: src/api/forms/components/FormProperties.vue [hunk 9]
```diff
@@ -105,6 +105,20 @@ export default {
...
+                return this.model.buttons.submit.label;
+            }
+
+            return 'OK';
+        },
+        cancelLabel() {
+            if (this.model.buttons && this.model.buttons.cancel && this.model.buttons.cancel.label) {
```
Comment (akhenry) [inline src/api/forms/components/FormProperties.vue:117]: This should be split over multiple lines

File: src/api/user/User.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    constructor(userId, name) {
+        this.userId = userId;
+        this.name = name;
+
+        this.getUserInfo = this.getUserInfo.bind(this);
+        this.getUserId = this.getUserId.bind(this);
+        this.getName = this.getName.bind(this);
+    }
+
+    getUserInfo() {
```
Comment (akhenry) [inline src/api/user/User.js:33]: Why does this method exist when we already have `getUserId()` and `getName()`?

**Review by akhenry — CHANGES_REQUESTED** [review]
Looks great. Just some small formatting suggestions and a question.

File: src/api/user/User.js (hunk 10 above)
Comment (jvigliotta) [reply src/api/user/User.js:33]: I think I just added it as a quick way to get all info. I can remove?

Comment (akhenry) [reply src/api/user/User.js:33]: Yes please. I want to keep the API as minimal as possible initially.

Comment (jvigliotta) [reply src/api/user/User.js]: Gotcha! Makes much more sense to return the user object, thanks!

**Review by akhenry — APPROVED** [review]
LGTM!

### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue) Y

**Pull request merged**